### ISSUE TITLE
🤖 Auto-Fix Bug-Report – 28.05.2026 – 4 Bugs

### ISSUE THREAD

**Issue by Henni2609** [Henni2609/Graph-RAG#2]
⚠️ Es existieren bereits offene auto-fix-pending Issues: #1 — die Auto-Fix-Routine arbeitet sie nacheinander ab.

## Auto-Fix Bug-Report

**Datum:** 28.05.2026
**Bugs gesamt:** 8
**Fix-relevant (Kritisch+Hoch+Mittel):** 4
**Verteilung:** Kritisch: 0, Hoch: 1, Mittel: 3, Niedrig: 4

---

## 📋 Zusammenfassung

Die Codebasis ist insgesamt sauber und gut strukturiert (sinnvolle Lock-Disziplin, durchgängiges Cypher-Parameter-Binding, klare Trennung der Pipeline-Phasen). Die Befunde sind überwiegend moderate Logik- bzw. Performance-Probleme. Der einzige Hoch-Befund ist ein DoS-Vektor im Upload-Endpoint, der die volle Datei ins RAM lädt, **bevor** die Größenprüfung greift.

### Bug #1: Upload liest volle Datei vor Größencheck → DoS-Risiko
- **Datei:** `src/kg_rag/web/app.py:373-381`
- **Schweregrad:** Hoch
- **Kategorie:** Security / DoS
- **Problem:** In `_enqueue_upload` wird `contents = await uf.read()` vor der Prüfung `len(contents) > MAX_UPLOAD_BYTES` ausgeführt. Damit wird eine z.B. 5-GB-Datei vollständig (Spool: erste 1 MB im RAM, Rest im tmp-File) gelesen, bevor sie abgelehnt wird. Mehrere parallele Uploads können RAM/Disk erschöpfen.
- **Auswirkung:** Ein einziger Client kann mit großen Multipart-Bodies den Server-Host (RAM/tmpfs) füllen. Auch ohne böse Absicht verschwenden zu große PDFs Zeit/Bandbreite, die sonst früh hätte abgelehnt werden können.
- **Lösungsplan:**
  1. `Content-Length`-Header vor `read()` prüfen und Anfrage frühzeitig mit 413 ablehnen.
  2. Zusätzlich `read()` in Chunks (`while chunk := await uf.read(1024*1024)`) und Größe inkrementell prüfen — abbrechen, sobald `MAX_UPLOAD_BYTES` überschritten.
  3. Optional: globalen Concurrency-Limit-Token für parallele Uploads.
- **Code-Beispiel (Fix):**

  **Vorher:**
  ```python
  contents = await uf.read()
  if not contents:
      raise HTTPException(status_code=400, detail=f"Datei ist leer: {filename}")
  if len(contents) > MAX_UPLOAD_BYTES:
      raise HTTPException(status_code=413, detail=f"Datei zu groß (max. 50 MB): {filename}")
  ```

  **Nachher:**
  ```python
  buf = bytearray()
  while chunk := await uf.read(1024 * 1024):
      buf.extend(chunk)
      if len(buf) > MAX_UPLOAD_BYTES:
          raise HTTPException(status_code=413, detail=f"Datei zu groß (max. 50 MB): {filename}")
  if not buf:
      raise HTTPException(status_code=400, detail=f"Datei ist leer: {filename}")
  contents = bytes(buf)
  ```
- **Geschätzter Aufwand:** 30 Min

---

### Bug #2: `section_title_search` gruppiert dokumentübergreifend → fehlende Sections bei Multi-PDF-Sessions
- **Datei:** `src/kg_rag/neo4j_store.py:402`
- **Schweregrad:** Mittel
- **Kategorie:** Logikfehler / Retrieval
- **Problem:** Cypher gruppiert mit `WITH c.section_title AS section_title, collect(c)[0] AS fc` nur auf `section_title`, nicht auf `(document_id, section_title)`. Haben mehrere PDFs in derselben Session jeweils eine „Conclusion"/„Schlussfolgerung", wird nur **eine** zurückgegeben (das Chunk mit dem global kleinsten `chunk_index`).
- **Auswirkung:** Frage „Was sind die Schlussbetrachtungen?" liefert nur die Conclusion eines Dokuments statt aller. Wirkt wie fehlende Information, obwohl indexiert.
- **Lösungsplan:**
  1. Den `WITH`-Schlüssel auf `(document_id, section_title)` erweitern.
  2. `LIMIT` ggf. anheben oder durch Per-Doc-Limit (Top-N pro Dokument) ersetzen.
- **Code-Beispiel (Fix):**

  **Vorher:**
  ```cypher
  WITH c
  ORDER BY c.chunk_index ASC
  WITH c.section_title AS section_title, collect(c)[0] AS fc
  ```

  **Nachher:**
  ```cypher
  WITH c
  ORDER BY c.chunk_index ASC
  WITH c.document_id AS doc_id, c.section_title AS section_title, collect(c)[0] AS fc
  ```
- **Geschätzter Aufwand:** 5 Min

---

### Bug #3: `_PDF_PAGES_CACHE` wächst unbegrenzt → Memory-Leak
- **Datei:** `src/kg_rag/web/app.py:113`
- **Schweregrad:** Mittel
- **Kategorie:** Memory / Resource Leak
- **Problem:** Das modul-globale Dict `_PDF_PAGES_CACHE` wird nur beim expliziten `DELETE /api/document/...` oder `POST /api/session/end` evictiert. Schließt der Nutzer einfach den Tab, bleiben die geparsten Seiten (inkl. OCR-Text — oft >1 MB pro PDF) für die gesamte Prozesslaufzeit im Speicher. Außerdem gibt es keine Obergrenze.
- **Auswirkung:** Bei mehreren Sessions/Tag schwillt der Python-Prozess kontinuierlich an, bis OOM-Kill oder Container-Limit.
- **Lösungsplan:**
  1. Cache durch `cachetools.TTLCache` oder eigene LRU mit Max-Size/Idle-TTL ersetzen.
  2. Periodischer Cleanup-Thread, der Einträge älter als z.B. `JOB_TTL_SECONDS` verwirft.
  3. Alternativ: an `JobState.finished_at + grace` koppeln.
- **Code-Beispiel (Fix-Skizze):**
  ```python
  from cachetools import TTLCache
  _PDF_PAGES_CACHE: TTLCache = TTLCache(maxsize=100, ttl=JOB_TTL_SECONDS)
  ```
- **Geschätzter Aufwand:** 30 Min

---

### Bug #4 (Verdacht): `vector_search` umgeht den erstellten Vector-Index → O(N)-Scan pro Query
- **Datei:** `src/kg_rag/neo4j_store.py:420-451`
- **Schweregrad:** Mittel (Verdacht — Code-Kommentar deutet auf bewusste Entscheidung)
- **Kategorie:** Performance
- **Problem:** `setup_schema` legt einen `VECTOR INDEX chunk_embeddings` an, aber `vector_search` nutzt ihn nicht. Statt `CALL db.index.vector.queryNodes(...)` läuft ein voller `MATCH (c:Chunk) WHERE c.session_id = $sid` über alle Chunks der Session und berechnet exakte Cosinus. Bei großen Multi-PDF-Sessions (>5k Chunks) wird das spürbar langsam.
- **Auswirkung:** Query-Latenz wächst linear mit Sessions-Korpus. Vector-Index ist toter Speicher.
- **Lösungsplan:**
  1. Hybridansatz: `queryNodes(index, k*4, $embedding)` mit Session-Postfilter, Fallback auf exakten Pfad wenn Treffer < k.
  2. Oder pro Session eigenen Index (Neo4j 5.18+: Multi-Index nach session_id-Label).
  3. Falls bewusst exakt bleiben soll: Kommentar präzisieren und Index-Erstellung entfernen, um falsche Erwartungen zu vermeiden.
- **Code-Beispiel (Fix-Skizze):**
  ```cypher
  CALL db.index.vector.queryNodes('chunk_embeddings', $top_k * 4, $embedding) YIELD node, score
  WHERE node.session_id = $session_id
  RETURN node.id AS id, ..., score
  ORDER BY score DESC LIMIT $top_k
  ```
- **Geschätzter Aufwand:** 1–2h (inkl. Tests + Recall/Latenz-Verifikation)

---

### Bug #5: BM25-Treffer werden im Kontext als „Via Graph-Traversal" gelabelt
- **Datei:** `src/kg_rag/components/context_merger.py:96-103`
- **Schweregrad:** Niedrig
- **Kategorie:** Logikfehler / UX
- **Problem:** BM25-Retriever setzt `retrieval_source="bm25"`. In `format_context` gibt es Branches nur für `"vector"`, `"section"` und einen `else`-Pfad „Via Graph-Traversal". BM25-only-Treffer fallen in den `else`-Zweig und werden als Graph-Treffer ausgewiesen — fachlich falsch.
- **Auswirkung:** Falsche Quellenbeschreibung im UI, beeinträchtigt Transparenz und Debugging.
- **Lösungsplan:**
  1. Expliziten Branch `elif retrieval_source == "bm25": label = "Lexikalisch relevant"` einfügen.
  2. Optional „graph" eindeutig matchen (`elif retrieval_source == "graph":`) und `else` zum echten Fallback machen.
- **Code-Beispiel (Fix):**
  ```python
  if retrieval_source == "vector":
      label = "Semantisch relevant"
  elif retrieval_source == "section":
      label = "Direkt aus Abschnitt"
  elif retrieval_source == "bm25":
      label = "Lexikalisch relevant"
  else:
      label = "Via Graph-Traversal"
  ```
- **Geschätzter Aufwand:** 5 Min

---

### Bug #6: `_resolve_device` unterstützt kein CUDA
- **Datei:** `src/kg_rag/pipelines/indexing.py:234-241`
- **Schweregrad:** Niedrig
- **Kategorie:** Logikfehler / Performance
- **Problem:** Auto-Erkennung prüft nur `torch.backends.mps.is_available()` (Apple Silicon). Auf Linux mit CUDA-GPU fällt das System auf CPU zurück, obwohl `torch.cuda.is_available()` True wäre.
- **Auswirkung:** Indexierung/Reranking auf GPU-Hosts läuft unnötig langsam.
- **Lösungsplan:** CUDA-Check vor MPS-Check hinzufügen.
- **Code-Beispiel (Fix):**
  ```python
  try:
      import torch
      if torch.cuda.is_available():
          return "cuda"
      if torch.backends.mps.is_available():
          return "mps"
      return "cpu"
  except Exception:
      return "cpu"
  ```
- **Geschätzter Aufwand:** 5 Min

---

### Bug #7: `_PDF_PAGES_CACHE` check-then-act-Race → doppelte Arbeit
- **Datei:** `src/kg_rag/web/app.py:221-232`
- **Schweregrad:** Niedrig
- **Kategorie:** Race Condition / Performance
- **Problem:** Lock wird zwischen Cache-Read und Cache-Write losgelassen. Zwei gleichzeitige Anfragen auf dasselbe Dokument extrahieren beide das PDF (teuer mit OCR), letzter Schreiber gewinnt.
- **Auswirkung:** Doppelte CPU-Last bei Klick-Spam o.ä.; keine Datenkorruption.
- **Lösungsplan:** Per-Document-Lock (z.B. `defaultdict(Lock)`) einsetzen oder Future-basierter Single-Flight.
- **Geschätzter Aufwand:** 30 Min

---

### Bug #8: `JobState.future` wird außerhalb des `JOBS_LOCK` zugewiesen
- **Datei:** `src/kg_rag/web/app.py:417-422`
- **Schweregrad:** Niedrig
- **Kategorie:** Race Condition
- **Problem:** `INDEXING_JOBS[job_id] = state` läuft unter Lock; `state.future = future` jedoch danach lockfrei. Aktuelle Leser nutzen `state.future` zwar nicht, aber der `_shutdown_jobs`-Handler mit `JOB_EXECUTOR.shutdown(cancel_futures=True)` könnte zukünftig Futures aus dem Dict lesen.
- **Auswirkung:** Aktuell latent; bei zukünftigen Änderungen NoneType-Fehler oder verlorene Cancel-Pfade.
- **Lösungsplan:** Future zuerst submitten, dann komplettes `JobState` mit Future unter Lock einsetzen — oder Zuweisung in den `with JOBS_LOCK`-Block aufnehmen.
- **Code-Beispiel (Fix):**
  ```python
  future = JOB_EXECUTOR.submit(_run_indexing_job, job_id, pdf_paths, config)
  with JOBS_LOCK:
      _evict_old_jobs(time.time())
      state.future = future
      INDEXING_JOBS[job_id] = state
  ```
- **Geschätzter Aufwand:** 5 Min

---

## 🎯 Priorisierte Roadmap

1. **Bug #1 (Hoch)** zuerst — DoS-Vektor in produktiv erreichbarem Endpoint, kleines Patch, hoher Sicherheitsgewinn.
2. **Bug #2 (Mittel)** danach — Trivialer Cypher-Fix mit direktem Retrieval-Quality-Gewinn bei mehreren PDFs.
3. **Bug #5 (Niedrig)** und **Bug #8 (Niedrig)** als Quick-Wins (je 5 Min) gleich mitnehmen.
4. **Bug #3 (Mittel)** — Memory-Leak vor längeren Production-Läufen beheben.
5. **Bug #6 (Niedrig)** — CUDA-Pfad, falls jemand auf GPU-Host deployt.
6. **Bug #7 (Niedrig)** — eher Optimierung, niedrige Priorität.
7. **Bug #4 (Mittel/Verdacht)** zuletzt: Entweder Cypher umbauen oder den Vector-Index entfernen + Kommentar präzisieren — sollte mit Recall/Latenz-Messung untermauert werden, daher höchster Aufwand.

---

_Automatisch erstellt von Claude Routine "Tägliches Code Review". Wird durch das Erstellen dieses Issues an die Routine "Auto-Fix Bug-Report" übergeben (GitHub-Trigger: "Issue eröffnet")._
